Source/WebCore/css/typedom/DeclaredStylePropertyMap.cpp

```diff
@@ -0,0 +1,122 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "DeclaredStylePropertyMap.h"
+
+#include "CSSStyleRule.h"
+#include "CSSStyleSheet.h"
+#include "StyleProperties.h"
+#include "StyleRule.h"
+
+namespace WebCore {
+
+Ref<DeclaredStylePropertyMap> DeclaredStylePropertyMap::create(CSSStyleRule& ownerRule)
+{
+    return adoptRef(*new DeclaredStylePropertyMap(ownerRule));
+}
+
+DeclaredStylePropertyMap::DeclaredStylePropertyMap(CSSStyleRule& ownerRule)
+    : m_ownerRule(ownerRule)
+{
+}
+
+unsigned DeclaredStylePropertyMap::size() const
+{
+    if (auto* styleRule = this->styleRule())
+        return styleRule->properties().propertyCount();
+    return 0;
+}
+
+auto DeclaredStylePropertyMap::entries(ScriptExecutionContext* context) const -> Vector<StylePropertyMapEntry>
+{
+    if (!context)
+        return { };
+
+    auto* styleRule = this->styleRule();
+    if (!styleRule)
+        return { };
+
+    auto& document = downcast<Document>(*context);
+    Vector<StylePropertyMapEntry> result;
+    auto& declaredStyleSet = styleRule->properties();
+    result.reserveInitialCapacity(declaredStyleSet.propertyCount());
+    for (unsigned i = 0; i < declaredStyleSet.propertyCount(); ++i) {
+        auto propertyReference = declaredStyleSet.propertyAt(i);
+        result.uncheckedAppend(makeKeyValuePair(propertyReference.cssName(), reifyValueToVector(propertyReference.value(), document)));
+    }
+    return result;
+}
+
+CSSValue* DeclaredStylePropertyMap::propertyValue(CSSPropertyID propertyID) const
+{
+    if (auto* styleRule = this->styleRule())
+        return styleRule->properties().getPropertyCSSValue(propertyID).get();
+    return nullptr;
+}
+
+CSSValue* DeclaredStylePropertyMap::customPropertyValue(const AtomString& propertyName) const
+{
+    if (auto* styleRule = this->styleRule())
+        return styleRule->properties().getCustomPropertyCSSValue(propertyName.string()).get();
```

Comment:
I prefer the style below where we early return the null case.

Reply:
I opted with the one that was less lines, but OK.